drivers/nvme.cc

```diff
@@ -0,0 +1,693 @@
+        return EIO;
+    }
+
+    _ns_data.insert(std::make_pair(ns, new nvme_ns_t));
+    _ns_data[ns]->blockcount = data->ncap;
+    _ns_data[ns]->blockshift = data->lbaf[data->flbas & 0xF].lbads;
+    _ns_data[ns]->blocksize = 1 << _ns_data[ns]->blockshift;
+    _ns_data[ns]->bpshift = NVME_PAGESHIFT - _ns_data[ns]->blockshift;
+    _ns_data[ns]->id = ns;
+    
+    trace_nvme_identify_namespace(ns, _ns_data[ns]->blockcount, _ns_data[ns]->blocksize);
+    return 0;
+}
+
+//identify all active namespaces with nsid >= start
+int nvme::identify_active_namespaces(u32 start)
+{
+    assert(start >= 1);
+    assert(_identify_controller);
+    //max number of namespaces supported by the controller
+    u32 nn = _identify_controller->nn;
+    assert(nn > start);
+
+    auto cmd = alloc_cmd();
+    cmd->identify.cns = 2;
+    cmd->identify.common.nsid = start - 1;
+    cmd->identify.common.opc = NVME_ACMD_IDENTIFY;
+    auto active_namespaces = (u64*) alloc_phys_contiguous_aligned(4096, 4);
+    memset(active_namespaces, 0, 4096);
+
+    _admin_queue->submit_and_return_on_completion(std::move(cmd), (void*) mmu::virt_to_phys(active_namespaces), 4096);
+    int err;
+    for(int i=0; i < 1024; i++) {
+        if(active_namespaces[i]) {
+            err = identify_namespace(active_namespaces[i]);
+            if (err) {
+                free_phys_contiguous_aligned(active_namespaces);
+                return err;
+            }
+        } else { break;}
+    }
+    free_phys_contiguous_aligned(active_namespaces);
+    return 0;
+}
+
+int nvme::make_request(bio* bio, u32 nsid)
+{  
+    if(bio->bio_bcount % _ns_data[nsid]->blocksize || bio->bio_offset % _ns_data[nsid]->blocksize) {
+        NVME_ERROR("bio request not block-aligned length=%d, offset=%d blocksize=%d\n",bio->bio_bcount, bio->bio_offset, _ns_data[nsid]->blocksize);
+        return EINVAL;
+    }
+    bio->bio_offset = bio->bio_offset >> _ns_data[nsid]->blockshift;
+    bio->bio_bcount = bio->bio_bcount >> _ns_data[nsid]->blockshift;
+
+    assert((bio->bio_offset + bio->bio_bcount) <= _ns_data[nsid]->blockcount);
+    
+    if(bio->bio_cmd == BIO_FLUSH && (_identify_controller->vwc == 0 || !NVME_VWC_ENABLED )) {
+        biodone(bio,true);
+        return 0;
+    }
+
+    if(sched::current_cpu->id >= _io_queues.size())
+        return _io_queues[0]->make_request(bio, nsid);
+
+    return _io_queues[sched::current_cpu->id]->make_request(bio, nsid);
+}
+
+void nvme::register_admin_interrupts() 
+{
+    sched::thread* aq_thread = sched::thread::make([this] { this->_admin_queue->req_done(); },
+        sched::thread::attr().name("nvme"+ std::to_string(_id)+"_aq_req_done"));
+    aq_thread->start();
+        
+    bool ok = msix_register(0, [this] { this->_admin_queue->disable_interrupts(); }, aq_thread);
+    _dev.msix_unmask_entry(0);
+    if(not ok)
+        printf("admin interrupt registration failed\n");
+}
+
+bool nvme::msix_register(unsigned iv,
```

Comment:
This code seems to have a lot of overlap with `arch/x64/msi.cc`. Any reason not to re-use it?